File: Backend/Controllers/UsersController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using MedicalSystem.Data;
using MedicalSystem.DTOs;
using MedicalSystem.Models;
using MedicalSystem.Helpers;

namespace MedicalSystem.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    [Authorize]
    public class UsersController : ControllerBase
    {
        private readonly ApplicationDbContext _context;

        public UsersController(ApplicationDbContext context)
        {
            _context = context;
        }

        // GET: api/users?search=&role=&page=1&pageSize=10
        [HttpGet]
        [Authorize(Roles = "Admin,Doctor,LabTechnician,Radiologist,Pharmacist,Receptionist,Cashier")]
        public async Task<IActionResult> GetAll([FromQuery] string? search, [FromQuery] string? role, [FromQuery] int page = 1, [FromQuery] int pageSize = 10)
        {
            page = Math.Max(page, 1);
            pageSize = Math.Clamp(pageSize, 1, 100);

            var query = _context.Users.AsQueryable();

            if (!string.IsNullOrEmpty(search))
                query = query.Where(u => u.FullName.Contains(search) || u.Email.Contains(search) || (u.Phone != null && u.Phone.Contains(search)));

            if (!string.IsNullOrEmpty(role))
                query = query.Where(u => u.Role == role);

            var totalCount = await query.CountAsync();

            var users = await query
                .OrderByDescending(u => u.CreatedAt)
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .Select(u => new UserInfoDTO
                {
                    UserID = u.UserID,
                    FullName = u.FullName,
                    Email = u.Email,
                    Role = u.Role,
                    Phone = u.Phone,
                    IsActive = u.IsActive,
                    CreatedAt = u.CreatedAt,
                    AssignedTreasuryID = u.AssignedTreasuryID,
                    AssignedTreasuryNameAr = u.AssignedTreasury != null ? u.AssignedTreasury.TreasuryNameAr : null,
                    ProfileID = u.Role == "Doctor"
                        ? u.DoctorProfile!.DoctorID
                        : u.Role == "Patient"
                            ? u.PatientProfile!.PatientID
                            : (int?)null,
                    Specialty = u.DoctorProfile != null ? u.DoctorProfile.Specialty : null,
                    LicenseNumber = u.DoctorProfile != null ? u.DoctorProfile.LicenseNumber : null,
                    ConsultationFee = u.DoctorProfile != null ? u.DoctorProfile.ConsultationFee : 0,
                    BloodType = u.PatientProfile != null ? u.PatientProfile.BloodType : null,
                    Gender = u.PatientProfile != null ? u.PatientProfile.Gender : null,
                    DateOfBirth = u.PatientProfile != null ? u.PatientProfile.DateOfBirth : null,
                    RiskLevel = u.PatientProfile != null ? u.PatientProfile.RiskLevel : null
                })
                .ToListAsync();

            return Ok(new PaginatedResponse<UserInfoDTO>
            {
                Data = users,
                TotalCount = totalCount,
                Page = page,
                PageSize = pageSize
            });
        }

        // GET: api/users/5
        [HttpGet("{id}")]
        public async Task<IActionResult> GetById(int id)
        {
            var user = await _context.Users
                .Where(u => u.UserID == id)
                .Select(u => new
                {
                    u.UserID,
                    u.FullName,
                    u.Email,
                    u.Role,
                    u.Phone,
                    u.IsActive,
                    u.CreatedAt,
                    u.AssignedTreasuryID,
                    AssignedTreasuryNameAr = u.AssignedTreasury != null ? u.AssignedTreasury.TreasuryNameAr : null,
                    DoctorProfile = u.DoctorProfile != null ? new
                    {
                        u.DoctorProfile.DoctorID,
                        u.DoctorProfile.Specialty,
                        u.DoctorProfile.LicenseNumber,
                        u.DoctorProfile.EmergencyReady,
                        u.DoctorProfile.Bio,
                        u.DoctorProfile.ImageUrl,
                        u.DoctorProfile.ConsultationFee,
                        u.DoctorProfile.AvailableDays,
                        u.DoctorProfile.WorkStartTime,
                        u.DoctorProfile.WorkEndTime
                    } : null,
                    PatientProfile = u.PatientProfile != null ? new
                    {
                        u.PatientProfile.PatientID,
                        u.PatientProfile.BloodType,
                        u.PatientProfile.ChronicDiseases,
                        u.PatientProfile.Allergies,
                        u.PatientProfile.Gender,
                        u.PatientProfile.DateOfBirth
                    } : null
                })
                .FirstOrDefaultAsync();

            if (user == null)
                return NotFound(ApiResponse.Fail("المستخدم غير موجود"));

            var currentUserId = JwtHelper.GetUserIdFromClaims(User);
            var role = JwtHelper.GetUserRoleFromClaims(User);

            // Admin يرى أي مستخدم؛ المستخدم يرى نفسه فقط؛ باقي الأدوار ممنوعة
            if (role != "Admin" && currentUserId != id)
                return Forbid();

            return Ok(ApiResponse<object>.Ok(user));
        }

        // POST: api/users
        [HttpPost]
        [Authorize(Roles = "Admin")]
        public async Task<IActionResult> Create([FromBody] RegisterDTO dto)
        {
            if (string.IsNullOrWhiteSpace(dto.FullName) || string.IsNullOrWhiteSpace(dto.Email) || string.IsNullOrWhiteSpace(dto.Password))
                return BadRequest(ApiResponse.Fail("الاسم والبريد الإلكتروني وكلمة المرور مطلوبة."));

            var allowedRoles = new[] { "Admin", "Doctor", "Patient", "Pharmacist", "LabTechnician", "Radiologist", "Receptionist", "Cashier", "WarehouseKeeper", "Accountant" };
            if (string.IsNullOrWhiteSpace(dto.Role) || !allowedRoles.Contains(dto.Role))
                return BadRequest(ApiResponse.Fail("الدور المحدد غير صالح."));

            if (dto.Role == "Doctor" && dto.ConsultationFee < 0)
                return BadRequest(ApiResponse.Fail("رسوم الكشف لا يمكن أن تكون قيمة سالبة."));

            if (dto.Role == "Cashier")
            {
                if (!dto.AssignedTreasuryID.HasValue)
                    return BadRequest(ApiResponse.Fail("يجب تحديد الخزينة المخصصة للكاشير."));
                var treasury = await _context.Treasuries.FindAsync(dto.AssignedTreasuryID.Value);
                if (treasury == null || !treasury.IsActive)
                    return BadRequest(ApiResponse.Fail("الخزينة المخصصة للكاشير غير موجودة أو غير مفعّلة."));
            }

            var exists = await _context.Users.AnyAsync(u => u.Email == dto.Email);
            if (exists)
                return BadRequest(ApiResponse.Fail("البريد الإلكتروني مسجل مسبقاً"));

            var user = new User
            {
                FullName = dto.FullName,
                Email = dto.Email,
                Password = BCrypt.Net.BCrypt.HashPassword(dto.Password),
                Role = dto.Role,
                Phone = dto.Phone,
                AssignedTreasuryID = dto.Role == "Cashier" ? dto.AssignedTreasuryID : null,
                IsActive = true,
                CreatedAt = DateTime.Now
            };

            _context.Users.Add(user);
            await _context.SaveChangesAsync();

            if (dto.Role == "Doctor")
            {
                _context.DoctorProfiles.Add(new DoctorProfile
                {
                    UserID = user.UserID,
                    Specialty = dto.Specialty ?? "عام",
                    LicenseNumber = dto.LicenseNumber,
                    ConsultationFee = dto.ConsultationFee
                });
                await _context.SaveChangesAsync();
            }
            else if (dto.Role == "Patient")
            {
                _context.PatientProfiles.Add(new PatientProfile
                {
                    UserID = user.UserID,
                    BloodType = dto.BloodType,
                    Gender = dto.Gender,
                    DateOfBirth = dto.DateOfBirth,
                    FirstName = dto.FirstName,
                    FatherName = dto.FatherName,
                    GrandfatherName = dto.GrandfatherName,
                    FamilyName = dto.FamilyName,
                    FileNumber = await FileNumberHelper.GenerateNextAsync(_context)
                });
                await _context.SaveChangesAsync();
            }

            return Ok(ApiResponse.Ok("تم إنشاء المستخدم بنجاح"));
        }

        // PUT: api/users/5
        [HttpPut("{id}")]
        [Authorize(Roles = "Admin")]
        public async Task<IActionResult> Update(int id, [FromBody] RegisterDTO dto)
        {
            var user = await _context.Users.FindAsync(id);
            if (user == null)
                return NotFound(ApiResponse.Fail("المستخدم غير موجود"));

            if (string.IsNullOrWhiteSpace(dto.FullName) || string.IsNullOrWhiteSpace(dto.Email))
                return BadRequest(ApiResponse.Fail("الاسم والبريد الإلكتروني مطلوبان."));

            var emailTaken = await _context.Users.AnyAsync(u => u.Email == dto.Email && u.UserID != id);
            if (emailTaken)
                return BadRequest(ApiResponse.Fail("البريد الإلكتروني مسجل لمستخدم آخر"));

            user.FullName = dto.FullName;
            user.Email = dto.Email;
            user.Phone = dto.Phone;

            // تحديث الخزينة المخصصة للكاشير إن أُرسلت
            if (user.Role == "Cashier")
            {
                if (!dto.AssignedTreasuryID.HasValue)
                    return BadRequest(ApiResponse.Fail("يجب تحديد الخزينة المخصصة للكاشير."));
                var treasury = await _context.Treasuries.FindAsync(dto.AssignedTreasuryID.Value);
                if (treasury == null || !treasury.IsActive)
                    return BadRequest(ApiResponse.Fail("الخزينة المخصصة للكاشير غير موجودة أو غير مفعّلة."));
                user.AssignedTreasuryID = dto.AssignedTreasuryID.Value;
            }

            if (!string.IsNullOrEmpty(dto.Password))
                user.Password = BCrypt.Net.BCrypt.HashPassword(dto.Password);

            await _context.SaveChangesAsync();
            return Ok(ApiResponse.Ok("تم تحديث بيانات المستخدم بنجاح"));
        }

        // PUT: api/users/5/toggle-active
        [HttpPut("{id}/toggle-active")]
        [Authorize(Roles = "Admin")]
        public async Task<IActionResult> ToggleActive(int id)
        {
            var user = await _context.Users.FindAsync(id);
            if (user == null)
                return NotFound(ApiResponse.Fail("المستخدم غير موجود"));

            var currentUserId = JwtHelper.GetUserIdFromClaims(User);
            if (user.UserID == currentUserId)
                return BadRequest(ApiResponse.Fail("لا يمكنك تعطيل حسابك الخاص."));

            user.IsActive = !user.IsActive;
            await _context.SaveChangesAsync();

            var msg = user.IsActive ? "تم تفعيل الحساب" : "تم تعطيل الحساب";
            return Ok(ApiResponse.Ok(msg));
        }

        // DELETE: api/users/5
        [HttpDelete("{id}")]
        [Authorize(Roles = "Admin")]
        public async Task<IActionResult> Delete(int id)
        {
            var user = await _context.Users
                .Include(u => u.DoctorProfile)
                .Include(u => u.PatientProfile)
                .FirstOrDefaultAsync(u => u.UserID == id);

            if (user == null)
                return NotFound(ApiResponse.Fail("المستخدم غير موجود"));

            _context.Users.Remove(user);
            try
            {
                await _context.SaveChangesAsync();
            }
            catch (DbUpdateException)
            {
                return BadRequest(ApiResponse.Fail("لا يمكن حذف المستخدم لوجود سجلات مرتبطة به (مواعيد/فواتير/فحوصات). يُفضّل تعطيل الحساب بدلاً من الحذف."));
            }
            return Ok(ApiResponse.Ok("تم حذف المستخدم بنجاح"));
        }

        // GET: api/users/stats
        [HttpGet("stats")]
        [Authorize(Roles = "Admin")]
        public async Task<IActionResult> GetStats()
        {
            var stats = new
            {
                TotalUsers = await _context.Users.CountAsync(),
                ActiveUsers = await _context.Users.CountAsync(u => u.IsActive),
                Doctors = await _context.Users.CountAsync(u => u.Role == "Doctor"),
                Patients = await _context.Users.CountAsync(u => u.Role == "Patient"),
                Admins = await _context.Users.CountAsync(u => u.Role == "Admin")
            };

            return Ok(ApiResponse<object>.Ok(stats));
        }
    }
}

File: Backend/wwwroot/app/pages/users/users.controller.js
(function() {
    'use strict';

    angular.module('medicalApp').controller('UsersController', UsersController);

    UsersController.$inject = ['UsersService', 'TreasuryService', 'toastr'];

    function UsersController(UsersService, TreasuryService, toastr) {
        var vm = this;

        vm.users = [];
        vm.userStats = {};
        vm.loading = true;
        vm.saving = false;
        vm.searchQuery = '';
        vm.filterRole = '';
        vm.page = 1;
        vm.pageSize = 10;
        vm.totalCount = 0;
        vm.totalPages = 0;

        // Modals
        vm.showAddModal = false;
        vm.showDeleteModal = false;
        vm.showProfileModal = false;
        vm.editingUser = null;
        vm.selectedUser = null;
        vm.selectedProfileUser = null;
        vm.newUser = getEmptyUser();

        // Functions
        vm.loadUsers = loadUsers;
        vm.saveUser = saveUser;
        vm.editUser = editUser;
        vm.deleteUser = deleteUser;
        vm.confirmDelete = confirmDelete;
        vm.toggleActive = toggleActive;
        vm.viewUserProfile = viewUserProfile;
        vm.getRoleAr = getRoleAr;

        activate();

        function activate() {
            loadUsers();
            loadStats();
            loadTreasuries();
        }

        function loadTreasuries() {
            TreasuryService.getTreasuries().then(function(res) {
                vm.treasuries = (res && res.data) ? res.data : [];
            }).catch(function() {
                vm.treasuries = [];
            });
        }

        function loadUsers() {
            vm.loading = true;
            UsersService.getUsers({
                search: vm.searchQuery,
                role: vm.filterRole,
                page: 1,
                pageSize: 500
            }).then(function(response) {
                vm.users = response.data;
                vm.totalCount = response.totalCount;
                vm.totalPages = response.totalPages;
            }).catch(function() {
                toastr.error('حدث خطأ في تحميل البيانات');
            }).finally(function() {
                vm.loading = false;
            });
        }

        function loadStats() {
            UsersService.getUserStats().then(function(response) {
                if (response.success) vm.userStats = response.data;
            });
        }

        function saveUser() {
            vm.saving = true;
            var promise;

            if (vm.editingUser) {
                promise = UsersService.updateUser(vm.editingUser.userID, vm.newUser);
            } else {
                promise = UsersService.createUser(vm.newUser);
            }

            promise.then(function(response) {
                if (response.success) {
                    toastr.success(response.message);
                    vm.showAddModal = false;
                    vm.newUser = getEmptyUser();
                    vm.editingUser = null;
                    loadUsers();
                    loadStats();
                } else {
                    toastr.error(response.message);
                }
            }).catch(function(err) {
                toastr.error(err.data ? err.data.message : 'حدث خطأ');
            }).finally(function() {
                vm.saving = false;
            });
        }

        function editUser(user) {
            vm.editingUser = user;
            vm.newUser = {
                fullName: user.fullName || user.FullName || '',
                email: user.email || user.Email || '',
                phone: user.phone || user.Phone || '',
                role: user.role || user.Role || 'Patient',
                assignedTreasuryID: user.assignedTreasuryID || null,
                password: '',
                specialty: user.specialty || user.Specialty || (user.doctorProfile ? user.doctorProfile.specialty : ''),
                licenseNumber: user.licenseNumber || user.LicenseNumber || (user.doctorProfile ? user.doctorProfile.licenseNumber : ''),
                consultationFee: user.consultationFee || user.ConsultationFee || (user.doctorProfile ? user.doctorProfile.consultationFee : 0),
                bloodType: user.bloodType || user.BloodType || (user.patientProfile ? user.patientProfile.bloodType : ''),
                gender: user.gender || user.Gender || (user.patientProfile ? user.patientProfile.gender : '')
            };
            vm.showAddModal = true;
        }

        function deleteUser(user) {
            vm.selectedUser = user;
            vm.showDeleteModal = true;
        }

        function confirmDelete() {
            UsersService.deleteUser(vm.selectedUser.userID).then(function(response) {
                if (response.success) {
                    toastr.success(response.message);
                    vm.showDeleteModal = false;
                    loadUsers();
                    loadStats();
                }
            }).catch(function() {
                toastr.error('حدث خطأ في حذف المستخدم');
            });
        }

        function toggleActive(user) {
            UsersService.toggleActive(user.userID).then(function(response) {
                if (response.success) {
                    toastr.success(response.message);
                    loadUsers();
                }
            });
        }

        function viewUserProfile(user) {
            vm.selectedProfileUser = user;
            vm.showProfileModal = true;
        }

        function getRoleAr(role) {
            var roles = { 
                'Admin': 'مدير', 
                'Accountant': 'محاسب',
                'Doctor': 'طبيب', 
                'Patient': 'مريض',
                'Pharmacist': 'صيدلاني',
                'LabTechnician': 'فني مختبرات',
                'Radiologist': 'أخصائي أشعة',
                'Receptionist': 'موظف استقبال',
                'Cashier': 'كاشير',
                'WarehouseKeeper': 'أمين مخزن'
            };
            return roles[role] || role;
        }

        function getEmptyUser() {
            return { fullName: '', email: '', password: '', phone: '', role: 'Patient', assignedTreasuryID: null, specialty: '', licenseNumber: '', consultationFee: 0, bloodType: '', gender: '' };
        }
    }
})();
